pkl-core/src/main/java/org/pkl/core/runtime/AnsiCodingStringBuilder.java

```diff
@@ -0,0 +1,277 @@
+   *   sb.append(AnsiCode.BOLD, "hello");
+   * });
+   *
+   * }</pre>
+   * </ul>
+   */
+  public AnsiCodingStringBuilder append(AnsiCode code, Runnable runnable) {
+    if (!usingColor) {
+      runnable.run();
+      return this;
+    }
+    var prevDeclaredCodes = declaredCodes;
+    declaredCodes = EnumSet.of(code);
+    declaredCodes.addAll(prevDeclaredCodes);
+    runnable.run();
+    declaredCodes = prevDeclaredCodes;
+    return this;
+  }
+
+  /**
+   * Append a string whose contents are unknown, and might contain ANSI color codes.
+   *
+   * <p>Always add a reset and re-apply all colors after appending the string.
+   */
+  public AnsiCodingStringBuilder appendUntrusted(String value) {
+    appendCodes();
+    builder.append(value);
+    if (usingColor) {
+      doReset();
+      doAppendCodes(currentCodes);
+    }
+    return this;
+  }
+
+  /**
+   * Append {@code value} to the string.
+   *
+   * <p>If called within {@link #append(AnsiCode, Runnable)}, applies any styles in the current
+   * context.
+   */
+  public AnsiCodingStringBuilder append(String value) {
+    appendCodes();
+    builder.append(value);
+    return this;
+  }
+
+  /**
+   * Append the string representation of {@code value} to the string.
+   *
+   * <p>If called within {@link #append(AnsiCode, Runnable)}, applies any styles in the current
+   * context.
+   */
+  public AnsiCodingStringBuilder append(char value) {
+    appendCodes();
+    builder.append(value);
+    return this;
+  }
+
+  /**
+   * Append the string representation of {@code value} to the string.
+   *
+   * <p>If called within {@link #append(AnsiCode, Runnable)}, applies any styles in the current
+   * context.
+   */
+  public AnsiCodingStringBuilder append(int value) {
+    appendCodes();
+    builder.append(value);
+    return this;
+  }
+
+  /**
+   * Append the string representation of {@code value} to the string.
+   *
+   * <p>If called within {@link #append(AnsiCode, Runnable)}, applies any styles in the current
+   * context.
+   */
+  public AnsiCodingStringBuilder append(Object value) {
+    appendCodes();
+    builder.append(value);
+    return this;
+  }
+
+  /** Returns a fresh instance of this string builder. */
+  public AnsiCodingStringBuilder newInstance() {
+    return new AnsiCodingStringBuilder(usingColor);
+  }
+
+  public PrintWriter toPrintWriter() {
+    return new PrintWriter(new StringBuilderWriter(builder));
+  }
+
+  /** Builds the data represented by this builder into a {@link String}. */
+  public String toString() {
+    // be a good citizen and unset any ansi escape codes currently set.
+    reset();
+    return builder.toString();
+  }
+
+  private void doAppendCodes(Set<AnsiCode> codes) {
+    if (codes.isEmpty()) return;
+    builder.append("\033[");
+    var isFirst = true;
+    for (var code : codes) {
+      if (isFirst) {
+        isFirst = false;
+      } else {
+        builder.append(';');
+      }
+      builder.append(code.value);
+    }
```

Comment:
```suggestion
    StringUtils.joinToStringBuilder(builder, codes, ";", (code) -> builder.append(code.value));
```

Reply:
I'd like to avoid a circular package dependency here (`utils` -> `runtime` -> `utils`), so I'm going to keep this one as-is.